### DISCUSSION TITLE
as4ClientUserMessage.sendMessageWithRetries can't delete attachment files

### DISCUSSION THREAD

**Discussion by TMain-Tomi** [phax/phase4#134] [Q&A] [answered]
Dear @phax :

I use AS4ClientUserMessage.addAttachment(file1, CMimeType.TEXT_PLAIN, null) to add a attachment and I want to delete files after I sent successful, But files can't be delete, it seems can't be delete after as4ClientUserMessage.sendMessageWithRetries.My code is as follows:
```
AS4ClientUserMessage as4ClientUserMessage = BaseDataPhase4.createMandatoryAttributesSuccessMessage(as4Partner)
for(File file1: fileList){
       if(file1.isFile()){
           as4ClientUserMessage.addAttachment(file1, CMimeType.TEXT_PLAIN, null)
         }
      }
as4ClientUserMessage = BaseDataPhase4.setKeyStore(as4ClientUserMessage, as4Partner)
as4ClientUserMessage.signingParams().setAlgorithmSign(ECryptoAlgorithmSign.getFromIDOrDefault(as4Partner.sign, ECryptoAlgorithmSign.SIGN_ALGORITHM_DEFAULT))
                                .setAlgorithmSignDigest (ECryptoAlgorithmSignDigest.getFromIDOrDefault(digest,ECryptoAlgorithmSignDigest.SIGN_DIGEST_ALGORITHM_DEFAULT));
as4ClientUserMessage.cryptParams ().setAlgorithmCrypt (ECryptoAlgorithmCrypt.getFromIDOrDefault(as4Partner.encrypt, ECryptoAlgorithmCrypt.ENCRPYTION_ALGORITHM_DEFAULT))
ResponseHandlerMicroDom responseHandlerMicroDom = new ResponseHandlerMicroDom()
IAS4OutgoingDumper dumper = AS4DumpManager.getOutgoingDumper()
AS4ClientSentMessage as4ClientSentMessage = as4ClientUserMessage.sendMessageWithRetries(as4Partner.protocolAddress, responseHandlerMicroDom, null, dumper,null);
 for(File file1: fileList){
       if(file1.isFile()){
        file1.delete()
            }
}
```
May you give me some advices?

Best Regards,
Tomi Lin

**Comment by phax** [comment, marked as answer]
The resource handling is a bit tricky, especially when it comes to build-in retries etc. in which case an `InputStream` must be repeatable.
The resolution to this problem is the usage of an `AS4ResourceHelper` that defines the maximum scope of "temporary resources". You pass in an instance of this class into the constructor of `AS4ClientUserMessage`. After closing the `AS4ResourceHelper` you can also delete the temporary files.
hth, Philip

**Reply by TMain-Tomi** [reply]
Dear @phax ,

Yes, files can delete after I close AS4ResourceHelper. Thanks for your help.

Best Regards,
Tomi Lin
